src/rust/engine/graph/src/tests.rs

```diff
@@ -531,6 +525,40 @@ async fn canceled_immediately() {
   );
 }
 
+#[tokio::test]
+async fn canceled_on_loss_of_interest() {
+  let _logger = env_logger::try_init();
+  let graph = Arc::new(Graph::new());
+
+  let delay_for_middle = Duration::from_millis(2000);
+  let start_time = Instant::now();
+  let context = {
+    let mut delays = HashMap::new();
+    delays.insert(TNode::new(1), delay_for_middle);
+    TContext::new(graph.clone()).with_delays(delays)
+  };
+
+  // Start a run, but cancel it well before the delayed middle node can complete.
```

Comment:
Instead of relying on on the timing to allow this test to run successfully, why not introduce a node that is the Node Graph equivalent of `futures::future::pending` (i.e., a Future that never resolves)? Then the node that is delaying for two seconds would just be a node that never completes.

Reply:
Possibly... the harness doesn't have infrastructure for that right now. A downside of using tasks that wait forever though is that when those tests fail they deadlock rather than failing with an error. So even when I do want to test that something "has not happened", I like to use a timeout to avoid hanging tests.

Reply:
Then just use this PR to have an annotation you can add to any test to have a timeout on it: https://github.com/tokio-rs/tokio/pull/1823

The Tokio maintainers declined to accept my PR, but it might be useful for this use case.

Reply:
But I agree with you, my comment is moot if the graph testing harness does not support indefinitely pending nodes.